src/compute/src/render/top_k.rs

```diff
@@ -317,6 +332,143 @@ where
+                        notificator.for_each(|time, _, _| {
+                            if let Some(aggs) = aggregates.remove(time.time()) {
+                                let mut session = output.session(&time);
+                                for (record_time, topk) in aggs {
+                                    session.give_iterator(topk.into_iter().map(
+                                        |(monoid, diff)| (monoid.row, record_time.clone(), diff),
+                                    ))
+                                }
+                            }
+                        });
+                    },
+                )
+                .as_collection()
+        }
+    }
+}
+
+/// Types for in-place intra-ts aggregation of monotonic streams.
+pub mod topk_agg {
+    use smallvec::SmallVec;
+
+    pub struct TopKBatch<T> {
+        updates: SmallVec<[(T, i64); 16]>,
+        clean: usize,
+        limit: i64,
+    }
+
+    impl<T: Ord> TopKBatch<T> {
+        pub fn new(limit: i64) -> Self {
+            Self {
+                updates: SmallVec::new(),
+                clean: 0,
+                limit,
+            }
+        }
+
+        /// Adds a new update, for `item` with `value`.
+        ///
+        /// This could be optimized to perform compaction when the number of "dirty" elements exceeds
+        /// half the length of the list, which would keep the total footprint within reasonable bounds
+        /// even under an arbitrary number of updates. This has a cost, and it isn't clear whether it
+        /// is worth paying without some experimentation.
+        #[inline]
+        pub fn update(&mut self, item: T, value: i64) {
+            self.updates.push((item, value));
+            self.maintain_bounds();
+        }
+
+        /// Compact the internal representation.
+        ///
+        /// This method sort `self.updates` and consolidates elements with equal item, discarding
+        /// any whose accumulation is zero. It is optimized to only do this if the number of dirty
+        /// elements is non-zero.
+        #[inline]
+        pub fn compact(&mut self) {
+            if self.clean < self.updates.len() && self.updates.len() > 1 {
+                self.updates.sort_by(|x, y| x.0.cmp(&y.0));
+                for i in 0..self.updates.len() - 1 {
+                    if self.updates[i].0 == self.updates[i + 1].0 {
+                        self.updates[i + 1].1 += self.updates[i].1;
+                        self.updates[i].1 = 0;
+                    }
+                }
+                let mut limit = self.limit;
+                self.updates.retain(|x| {
+                    if limit > 0 {
+                        limit -= x.1;
+                        true
+                    } else {
+                        false
+                    }
+                });
+                // Adjust the diff of the last record that was retained so that we have exactly K
```

Comment:
I think we might end up happier if this struct owns the `Vec<ColumnOrder>` rather than having `T` own that and having `T::cmp` run the logic. At least, it reduces the number of allocations from "one per record" down to "one per worker".

Reply:
I agree, the monotonic top1 rendering has the same problem. The easiest fix that will improve the situation is to change it to an `Arc<Vec<ColumnOrder>>` but that won't survive across exchanges. To go further we'd have to either not rely on pure `Ord` traits or do some kind of data interning.

What I wanted to experiment with is a data interning scheme where each worker deterministically builds a thread local map of `(dataflow_id, item_id) -> Rc<dyn Any>` so that operators can associate any piece of data that is common across dataflows at dataflow construction with a small amount of data that is what gets exchanged over the network. Has this pattern been explored before?

Reply:
I think elsewhere we don't rely on `Ord`, but that seems fine. There is a `mz_expr::compare_columns` we use to sort things elsewhere, so anywhere it says `sort` we could use `sort_by(that)` instead.